Updated canals to check if at least 1 is passable instead of all if

ron-murhammer opened the pull request:
multiple canals between territories

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/triplea-game/triplea/343)
<!-- Reviewable:end -->

ron-murhammer commented:
Actual functional changes start around line 1400. Everything else is just formatting.

DanVanAtta commented:
Broken build, please fix, or close the PR. Since this was  one-off test failure, please either fix the test, or document the (broken) non-deterministic test and then re-run the build until it passes.

DanVanAtta commented:
@ron-murhammer @veqryn , in the future, I intend to close PRs that have a broken build and are more than a few hours old. I expect you two to do the same.

ron-murhammer commented:
@DanVanAtta Build passed now. I believe you've already opened an issue regarding the unit test that rarely fails.

DanVanAtta commented:
Thanks for re-running the build.  Indeed we do have a ticket. Though, it's not always the same method that fails, so documenting which method would be helpful potentially. Regardless, documenting each PR build failure occurrence I think has value from both a tracking and a process point of view.

DanVanAtta commented:
Hmm, I'm not very happy with this PR. I think this is case where we finally *need* a regression test put in place. I have no confidence that this change did not cause a regression. That gets to the second, point, it's really hard to tell what has changed vs formatting and actual changes. I'd like to request you split the commit into two. Secondly, why was formatting actually necessary? I thought we mass formatted everything..

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
-          continue;
+      failureMessage = canPassThroughCanal(canalAttachment, units, player, data);
+      if (failureMessage == null) {
+        return null;
```
Do not return null, return Optional.empty(), please fix this.

DanVanAtta replied:
This looks wrong, why does the for loop not continue through each and every canal attachment? As is, it stops at only the first one that it does not need to check. If that is the truly desired behavior, please change the "needToCheck" method to be a "getNeedsToCheck", and remove the for loop entirely.

DanVanAtta replied:
Okay, I think I see the logic now. This needs a comment to say we are stopping on the first canal we find that we can pass through.

ron-murhammer replied:
So Optional.empty() would be better but I tried to limit the amount of changes I made to the original code which used strings for error message and null if there are no errors.

I think the code is pretty straightforward and seems like you quickly realized the functionality as well. Essentially if you can pass through a canal then return. The java doc comment also describes this. Not sure its worth adding more comments here.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
-      final CanalAttachment attachment = iter.next();
-      if (attachment == null) {
+    for (final CanalAttachment canalAttachment : canalAttachments) {
+      if (!needToCheckCanal(canalAttachment, route, data)) {
```
Having multiple commits for the refactor would have made the review easier. I take some objection since I feel a bit of a double standard, harsher comments in my PRs about things not belonging, while here we have everything in one commit..

DanVanAtta replied:
Line 1428 doesn't make sense. If we do not need to check canal, then we check if we can pass through the canal. Is 1428 supposed to have the "!" in the if statement?

ron-murhammer replied:
So needToCheckCanal determines if the canal needs to be checked based on if it is actually on the route. If it doesn't need checked then that means its irrelavant (not on the units route) so its skipped. This is due to how canal attachments are attached to territories.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
+  /**
+   * Checks if route is either null or includes both canal territories so needs to be checked.
+   */
+  private static boolean needToCheckCanal(final CanalAttachment canalAttachment, final Route route, final GameData data) {
```
I like that you broke up these methods. I somewhat wish you had made them protected and tested them as well.

ron-murhammer replied:
Fair point. The struggle here is the number of test cases that needs written to really test canals well is fairly large as I don't believe any canal unit tests exist currently.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
+    return failureMessage;
+  }
+
+  /**
```
I like to see private methods with non-javadoc comments "/*". Javadoc'ing private methods is an anti-pattern, adds a lot of unused maintenance overhead.

ron-murhammer replied:
Agree. I'll change them to non-java doc though doesn't make a huge difference since no one really reads are java docs and a good portion of even public methods have no java doc.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
-    while (iter.hasNext()) {
-      final CanalAttachment attachment = iter.next();
-      if (attachment == null) {
+    for (final CanalAttachment canalAttachment : canalAttachments) {
```
Looks like canalAttachments could be null, need an NPE check. Please fix.

ron-murhammer replied:
I don't believe they can be which is why I removed the NPE.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
+    }
+    Territory last = null;
+    final Set<Territory> connectionToCheck = CanalAttachment.getAllCanalSeaZones(canalAttachment.getCanalName(), data);
+    for (final Territory current : route.getAllTerritories()) {
```
Potential NPE if route.getAllTerritories() is null, please fix.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
+    final Set<Territory> connectionToCheck = CanalAttachment.getAllCanalSeaZones(canalAttachment.getCanalName(), data);
+    for (final Territory current : route.getAllTerritories()) {
+      if (last != null) {
+        final Collection<Territory> lastTwo = new ArrayList<Territory>();
```
Do not couple the class type between RHS and LHS of an assignment unnecessarily. Please use `new ArrayList<>` or `Lists.newArrayList()`, ,please fix.

DanVanAtta replied:
Please use `ImmutableSet.of( last, current)` instead of the immtuble Collection<Territory> object

DanVanAtta replied:
This for loop can be simplified.. I would opt to have this functionality tucked into "Route", effectively: `route.containsSeaCanal( CanalAttachment)`

At the very least add a comment that indicates we know "conenctionToCheck" will be of size 2. That fact really helps make this more clear.

ron-murhammer replied:
So I didn't do any changes to this code except move it into a separate helper method. I'd prefer not to refactor code that works fine now and is now pushed down into a helper method.

DanVanAtta replied:
Moving code around to a helper method is refactoring. I would more prefer to see refactoring be done completely and properly when it is done, or alternatively explicitly called out that is what you are doing and is where you are stopping. This really should have been 3 different commits.

DanVanAtta replied:
Change a Set<> to ImmutableSet is really safe, and will make it easier to read the code on subsequent passes. It's the same argument as splitting the methods up. Some basic things like that are time efficient while we have all context switched and are looking at this code

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
+    if (route == null) {
+      return true;
+    }
+    Territory last = null;
```
SRP violation. The variable 'last' has different meanings depending upon if it is null or not null. Second smell is the use of nulls at all, leads to NPE exceptions. Use Optional, or if you know an assignment will happen, do this: :`final Territory last;`

ron-murhammer replied:
So I didn't do any changes to this code except move it into a separate helper method. I'd prefer not to refactor code that works fine now and is now pushed down into a helper method.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
+  /**
+   * Checks if units can pass through canal and returns null if true or a failure message if false.
+   */
+  private static String canPassThroughCanal(final CanalAttachment canalAttachment, final Collection<Unit> units,
```
Return Optional instead of a String value that has a double meaning and potential NPE.

DanVanAtta replied:
Even better yet, I think this method should have been broken up a bit differently/further. We also have a SRP violation since this method checks many different things. I recommend this be fixed further by splitting this up into two or three methods that return simple "true/false" answers. Let the caller code then formulate the message that is displayed.

ron-murhammer replied:
You could argue that breaking it up further would be even better but I'd say this is a step in the right direction and I'd prefer not to refactor it any further at this point.

ron-murhammer replied:
Went ahead and changed to use Optional.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
+   * Checks if route is either null or includes both canal territories so needs to be checked.
+   */
+  private static boolean needToCheckCanal(final CanalAttachment canalAttachment, final Route route, final GameData data) {
+    if (route == null) {
```
We're calling a static method with a null arg? Seems like this is probably a bug in the caller code. If this code repeatedly gets called with all null values, sounds like that would be unexpected and we would get a silent "true" back. IMO, no arg check or an IllegalArg if this is null, not silently returning return.

ron-murhammer replied:
So this is due to how the parent method is defined as it allows a null route as a perfectly valid argument. Really just passing it down to the helper method so it doesn't need to be dealt with further up the chain.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1652,9 +1675,12 @@ public static Route getBestRoute(final Territory start, final Territory end, fin
-              : Matches.TerritoryIsImpassable));
+      defaultRoute =
+          data.getMap().getRoute_IgnoreEnd(
+              start,
```
Bad auto-formatting here, please fix..

ron-murhammer replied:
Fixed.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1675,17 +1699,18 @@ public static Route getBestRoute(final Territory start, final Territory end, fin
       if (isNeutralsImpassable) {
-        landRoute = data.getMap().getRoute_IgnoreEnd(start, end,
-            new CompositeMatchAnd<Territory>(Matches.TerritoryIsLand, noNeutral, noImpassible));
+        landRoute =
```
It seems odd this was re-formatted. Please let us know if there is something actually off with the check-in formater, if not, please fix this so it is not formatted so weird. Maybe by assigning the CompositeMatchAnd to an intermediary variable, could get it formatted nicely that way perhaps.

ron-murhammer replied:
Fixed.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1675,17 +1699,18 @@ public static Route getBestRoute(final Territory start, final Territory end, fin
       } else {
-        landRoute = data.getMap().getRoute_IgnoreEnd(start, end,
-            new CompositeMatchAnd<Territory>(Matches.TerritoryIsLand, noImpassible));
+        landRoute =
```
Again, weird formatting, an accumulation of more technical debt. Unless you're formatting thousands and thousands of lines, the rule I know is you're not allowed to let the auto-formatter degrade the existing format. IE: boyscout rule, code is left as clean or cleaner than when you found it.

ron-murhammer replied:
Fixed.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1675,17 +1699,18 @@ public static Route getBestRoute(final Territory start, final Territory end, fin
-              .getLargestMovementCost(unitsWhichAreNotBeingTransportedOrDependent))
-              || (forceLandOrSeaRoute
-                  && Match.someMatch(unitsWhichAreNotBeingTransportedOrDependent, Matches.UnitIsLand)))) {
+              .getLargestMovementCost(unitsWhichAreNotBeingTransportedOrDependent)) || (forceLandOrSeaRoute && Match
```
Again, really weird formatting..

ron-murhammer replied:
Fixed.

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1709,13 +1734,13 @@ public static Route getBestRoute(final Territory start, final Territory end, fin
     if (isNeutralsImpassable) {
       tests = new ArrayList<Match<Territory>>(Arrays.asList(
-          // best if no enemy and no neutral
+      // best if no enemy and no neutral
```
Looks like a tab was removed - good. Makes me wonder what we were doing with a tab in this file. Though, it does look like the indention is now off.

ron-murhammer replied:
Yeah, something was definitely off around the formatting of this file...

DanVanAtta commented on src/games/strategy/triplea/delegate/MoveValidator.java:
```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
    */
   public static String validateCanal(final Territory territory, final Route route, final Collection<Unit> units,
       final PlayerID player, final GameData data) {
+    String failureMessage = null;
```
Don't use nulls (unless there is a specific reason), use Optional instead.

ron-murhammer replied:
Changed to use optional.

ron-murhammer commented:
@DanVanAtta Addressed your comments with a few follow on commits. 

The formatting is kind of funky and not sure what the deal is with this class. I made sure I have the latest formatting profiles by re-importing. I didn't realize any formatting was changed until after I made the PR otherwise it would have been a separate commit. Guess I'll make sure to check moving forward.

More refactoring could be done to this but I think the higher level method validateCanal is pretty clean now. Any remaining messy code was pushed down to private methods that have pretty good names to indicate what they do.

DanVanAtta commented:
You're touching the code at a pretty detailed level, I think some of the other basic refactoring should be done while we are paying the context switch cost (during review and coding). I think some of the basics need to be covered, like not using nulls and diamond notation. 

I don't like hearing changes were leaked. I urge  you use git for windows command line option to open a shell and use `git diff` to double check what you are committing. When doing rebases and merge conflicts, I've made mistakes there and that's easy, but I don't think commit leaks should have happened here. It's worrisome.

Regardless of how the bad formatting happened, please fix it (so someone else doesn't have to suffer through it and/or fix it)

DanVanAtta commented:
Summarizing:
- Since you started refactoring, I'd like to see you take one more pass and fix the nulls that are passed around, if not spending 20 minutes more and potentially finishing the clean up 
- The for loop that checks for any canal needs to be clarified one way or another so it doesn't look like a bug. Thinking about that, I recommend extracting that entire for loop into a method of its own that returns a boolean. Then from there you can have a small comment in the helper method if needed stating we are only looking for the first canal that doesn't have a move problem.
- Second thought, IDC as much if you fix the formatting anymore.. I'd rather you spend time elsewhere honestly..

veqryn commented:
Did we decide against moving the formatting to a different PR?

I will try to review this later in the week....

ron-murhammer commented:
@DanVanAtta Not really following you here. The for loop is pretty straightforward and the java doc comment on the method describes it pretty well:
"Used for testing a single territory, either as part of a route, or just by itself. Returns Optional.empty if either there are no canals or there is at least 1 canal that can be passed through otherwise returns a failure message indicating why the canal can't be passed through."

I can add an additional line comment into the for loop reiterating "there is at least 1 canal that can be passed through" if you feel that is needed.

@veqryn Yeah, I left the formatting changes in this PR and would appreciate another set of eyes since we are touching delegate logic here. It appears the formatting mostly had to do with indentation that wasn't properly formatted. 

The only functional changes start around line 1400 with the method validateCanal being updated and refactored into validateCanal, isCanalOnRoute, and canPassThroughCanal. Then just updating the 2 references to validateCanal to handle Optional instead of null checks.

DanVanAtta commented:
@ron-murhammer , thanks for pointing out the javadoc and the updates you made. I think things are good enough now, the end goal has been achieved and we have a few very nice quality improvements along the way.

@veqryn  / @ron-murhammer , I expressed a strong desire for this to have multiple commits. Regardless, I don't see value in forcing @ron-murhammer  to redo this, so simply giving him some crap about it and doing the review anyways has been my strategy ;)

I'm happy with this being merged.

ron-murhammer commented:
@veqryn Gonna assign this to you for review.

veqryn commented:
I have only one question: why use `java.util.Optional`?

With the exception of the null check on attachments, I don't think any logic or implementation was changed beyond your goal of having an OR relationship between canals.  Therefore I give this a :+1:

ron-murhammer commented:
@veqryn Using optional makes the returning of 'null' more explicit and easier to manage. The general practice now with Java 8 is to return Optional if the return parameter can be null. This makes the callers to the method make sure to check whether a value was returned or not.

Merging this now so we can get it in front of some testers.
